### DISCUSSION TITLE
Why  I must  overload operator false?

### DISCUSSION THREAD

**Discussion by zms9110750** [dotnet/csharplang#7671] [Q&A] [answered]
Machine translation may cause difficulty in expression, and reading may require imagination.

I flipped through the document and saw that 'false' works like this
- If you overload  at the same time "&"
- And your "&" return type is yourself

So you can get  "&&"

This means that if I don't write & or if the return type is not my own, then false is of no use.
So, why must false and true be overloaded in pairs?

**Comment by CyrusNajmabadi** [comment]
Primarily because if you do *not* overload this, you could end up with very strangely behaving types/values.  This acts as a safeguard, forcing you to at least think about this so you don't accidentally get into a confusing situation.  

Given that it is generally trivial to overload the other at the same time, this protection is viewed as more useful than not.

**Reply by zms9110750** [reply]
I didn't understand. The function of true is to use if, while, and ternary operators as conditions.
He himself is not a `bool`. If you want to reverse it, you need to overload it separately `!` or `~`.
This only works with 'true'. What strange behavior will occur?

**Reply by CyrusNajmabadi** [reply, marked as answer]
If you overload operator `true` then usages in `||` will properly short circuit.  If you overload operator `false` then usages in `&&` will properly short circuit.  If you don't overload operator `false` then shortcircuiting will not work properly.  That could lead to confusion where `||` works as one expects, but `&&` does not.

**Reply by huoyaoyuan** [reply]
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/true-false-operators

The doc may need the update to clarify the short-circuiting more.

**Reply by zms9110750** [reply]
What I mean is that `false` must cooperate with `|` to take effect.
So why do I have to write `false` when I don't have `|`?
He is completely incapable of working in this situation.

**Reply by CyrusNajmabadi** [reply]
Honestly, because our view is that it's just sensible to provide these as a pair, to prevent confusion.  And that the number of people inconvenienced by this is practically nil.

**Comment by kyoyama-kazusa** [comment]
`operator true` and `operator false` are a pair of operators that makes the logical operators work. If we don't have one of those two, we cannot directly know which operator I can use and which I cannot. Although I may not use both of them, but this doesn't mean it is unnecessary to overload both operators `true` and `false`.

Just like `<` and `>`, `<=` and `>=`, `==` and `!=`, I can control the case which operators I don't use, but "both are overloaded" are more reasonable and more intuitive. The pair `true` and `false` should not be an exception - they are a pair of definitions that we cannot separate to learn.

From the other hand to explain this, `&` and `&&` are not same, and `|` and `||` as well. Just think about `operator &` and `operator |`. The cases using such operators are not only for logical cases, but also they can be used in bitwise operations (or even set operations, i.e. intersection or union). Bitwise operations will work with many integer numbers (the number of the operating values are not only limited to 2), set operations will work with numbers of elements stored in a collection (the number of operating elements are also not only limited to 2), but logical operations only serves with boolean literals `true` and `false` (which only holds 2 cases). In my personal opinion, making the compiler allow not both being overloaded is just like I only handles for one boolean case, and there will be a missing for the other one, which will make the whole boolean system to "be incomplete".
